### PR TITLE
[TensorFlow][Graviton][Build][Sagemaker] TF 2.9 Graviton Release

### PR DESCRIPTION
**Pull request by xncqr**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
Updating TF Graviton from 2.7 to 2.9 with an additional image for sagemaker

The Local SM PR testing will fail as the test host is not the correct arch. Will be doing another PR to update the tests for this support. The Local SM testing passed manually from an arm64 host.

Remote SM testing through PR will fail as the SM endpoint does now know what Graviton is. Image was manually verified through an endpoint with the support.

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [x] `sagemaker_local_tests = true`

## Label Checklist
- [x] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tensorflow/buildspec-graviton.yml [hunk 1]
```diff
@@ -33,6 +37,20 @@
       tag_python_version: &TAG_PYTHON_VERSION py38
       os_version: &OS_VERSION ubuntu20.04
       tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION, "-ec2" ]
-      docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile.graviton.ec2., *DEVICE_TYPE ]
+      docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile.graviton., *DEVICE_TYPE ]
```
Comment (kenny-ezirim) [inline tensorflow/buildspec-graviton.yml:40]: Any reason for leaving out ec2 here (i.e. /Dockerfile.graviton.) but having ec2 in buildspec-graviton-2.7.yml definition file?

File: tensorflow/buildspec-graviton.yml [hunk 2]
```diff
@@ -33,6 +37,20 @@
...
+    BuildSageMakerTensorflowCPUInferencePy3DockerImage:
+      <<: *INFERENCE_REPOSITORY
+      build: &TENSORFLOW_CPU_INFERENCE_PY3 false
+      image_size_baseline: 4899
+      device_type: &DEVICE_TYPE cpu
+      python_version: &DOCKER_PYTHON_VERSION py3
+      tag_python_version: &TAG_PYTHON_VERSION py38
+      os_version: &OS_VERSION ubuntu20.04
+      tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION, "-sagemaker" ]
+      docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile.graviton., *DEVICE_TYPE ]
```
Comment (kenny-ezirim) [inline tensorflow/buildspec-graviton.yml:52]: Did we skip ec2 here because this is Sagemaker image?

File: tensorflow/buildspec-graviton-2.7.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    BuildEC2TensorflowCPUInferencePy3DockerImage:
+      <<: *INFERENCE_REPOSITORY
+      build: &TENSORFLOW_CPU_INFERENCE_PY3 false
+      image_size_baseline: 5000
+      device_type: &DEVICE_TYPE cpu
+      python_version: &DOCKER_PYTHON_VERSION py3
+      tag_python_version: &TAG_PYTHON_VERSION py38
+      os_version: &OS_VERSION ubuntu20.04
+      tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION, "-ec2" ]
+      docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile.graviton.ec2., *DEVICE_TYPE ]
```
Comment (kenny-ezirim) [inline tensorflow/buildspec-graviton-2.7.yml:37]: Additional label ec2 added here.

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py [hunk 4]
```diff
@@ -189,14 +189,14 @@ def train_mnist_model(serving_folder_path, ec2_connection, python_invoker):
 def host_setup_for_tensorflow_inference(
     serving_folder_path, framework_version, ec2_connection, is_neuron, is_graviton, model_name, python_invoker
 ):
-    # Installing protobuf at 3.20.* to fix errors in using latest protobuf
-    ec2_connection.run((f"{python_invoker} -m pip install --user -qq -U 'protobuf>=3.20,<3.21'"), hide=True)
+    # Install PIP so we can test
+    ec2_connection.run((f"sudo apt-get update && sudo apt-get install -y python3-pip"), hide=True)
```
Comment (kenny-ezirim) [inline test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:193]: Does this fix handle the installation of the right version of protobuf?

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py [hunk 5]
```diff
@@ -189,14 +189,14 @@ def train_mnist_model(serving_folder_path, ec2_connection, python_invoker):
...
 
-    # Tensorflow 1.x doesn't have package with version 1.15.2 so use only 1.15
+    # Per graviton team, need to install tensorflow from pypi with no version pinning. attempting a pin will result is
+    # pip not finding the version. the internal repo only has a custom Tensorflow 2.6 which is not compatible with TF 2.9+
+    # and this was the recommended action by the Graviton team.
     if is_graviton:
-        # TF training binary is used that is compatible for graviton instance type
-        TF_URL = "https://aws-dlc-graviton-training-binaries.s3.us-west-2.amazonaws.com/tensorflow/2.6.0/tensorflow-2.6.0-cp38-cp38-linux_aarch64.whl"
-        ec2_connection.run((f"{python_invoker} -m pip install --no-cache-dir -U {TF_URL}"), hide=True)
+        ec2_connection.run((f"{python_invoker} -m pip install --no-cache-dir -U tensorflow-cpu-aws"), hide=True)
```
Comment (kenny-ezirim) [inline test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:199]: Is tensorflow-cpu-aws a valid URI?

File: tensorflow/buildspec-graviton.yml (hunk 2 above)
Comment (xncqr) [reply tensorflow/buildspec-graviton.yml:52]: the EC2 is above. This is a sagemaker image. there are 2 images in this file.

File: tensorflow/buildspec-graviton-2.7.yml (hunk 3 above)
Comment (xncqr) [reply tensorflow/buildspec-graviton-2.7.yml:37]: That is the old image buildspec and should be there. this insure we can build old for patching. OS version will be different. thus this yml is "2.7" and the other has no version because it's the current buildspec

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py (hunk 4 above)
Comment (xncqr) [reply test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:193]: yes as this installs pip on the testing host, allowing for protobuf to be installed on the testing host. mind, this is not the container, but the host which is testing the container.

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py (hunk 5 above)
Comment (xncqr) [reply test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:199]: it's not a uri.. it's an official pypi package that is required for the new TF2.9 testing. Worked this with graviton team on this one and that is the needed package for the test.

File: tensorflow/buildspec-graviton.yml (hunk 1 above)
Comment (xncqr) [reply tensorflow/buildspec-graviton.yml:40]: because that does not exist in that folder. Dockerfile.graviton.yaml has both EC2 and SM images in it.

**Review by kenny-ezirim — DISMISSED** [review]

File: testspec.yml [hunk 6]
```diff
@@ -8,6 +8,10 @@ phases:
   pre_build:
     commands:
       - start-dockerd
+      - |
+        if expr "${ARCH_TYPE}" : "graviton" >/dev/null; then
+          docker run --rm --privileged multiarch/qemu-user-static --reset -p yes
```
Comment (saimidu) [inline testspec.yml:13]: Please add a comment on the buildspec or here in this comment thread about why this statement is necessary.

File: tensorflow/buildspec-graviton-2.7.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
+  account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
```
[the remark this answers is not in the document]
Comment (xncqr) [reply tensorflow/buildspec-graviton-2.7.yml]: fixed

File: tensorflow/buildspec-graviton.yml [hunk 8]
```diff
@@ -1,8 +1,8 @@
   account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
```
[the remark this answers is not in the document]
Comment (xncqr) [reply tensorflow/buildspec-graviton.yml]: fixed

File: tensorflow/buildspec-graviton.yml [hunk 9]
```diff
@@ -33,6 +37,20 @@
...
+      device_type: &DEVICE_TYPE cpu
+      python_version: &DOCKER_PYTHON_VERSION py3
+      tag_python_version: &TAG_PYTHON_VERSION py38
+      os_version: &OS_VERSION ubuntu20.04
+      tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION, "-sagemaker" ]
+      docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile.graviton., *DEVICE_TYPE ]
+      target: sagemaker
+      enable_test_promotion: true
+      context:
+        <<: *INFERENCE_CONTEXT
```
[the remark this answers is not in the document]
Comment (xncqr) [reply tensorflow/buildspec-graviton.yml]: fixed

File: tensorflow/inference/docker/2.9/py3/Dockerfile.graviton.cpu [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+ && curl -o ${HOME_DIR}/oss_compliance.zip https://aws-dlinfra-utilities.s3.amazonaws.com/oss_compliance.zip \
+ && unzip ${HOME_DIR}/oss_compliance.zip -d ${HOME_DIR}/ \
+ && cp ${HOME_DIR}/oss_compliance/test/testOSSCompliance /usr/local/bin/testOSSCompliance \
+ && chmod +x /usr/local/bin/testOSSCompliance \
+ && chmod +x ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh \
+ && ${HOME_DIR}/oss_compliance/generate_oss_compliance.sh ${HOME_DIR} ${PYTHON} \
+ && cp ${HOME_DIR}/oss_compliance/build_from_source_packages/BUILD_FROM_SOURCE_PACKAGES_LICENCES_AARCH64_IMAGES ${HOME_DIR} \
+ && rm -rf ${HOME_DIR}/oss_compliance*
+
+RUN curl https://aws-dlc-licenses.s3.amazonaws.com/tensorflow-2.7/license.txt -o /license.txt
```
[the remark this answers is not in the document]
Comment (xncqr) [reply tensorflow/inference/docker/2.9/py3/Dockerfile.graviton.cpu]: License updated.

File: tensorflow/buildspec-graviton-2.7.yml [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+      build: &TENSORFLOW_CPU_INFERENCE_PY3 false
+      image_size_baseline: 5000
+      device_type: &DEVICE_TYPE cpu
+      python_version: &DOCKER_PYTHON_VERSION py3
+      tag_python_version: &TAG_PYTHON_VERSION py38
+      os_version: &OS_VERSION ubuntu20.04
+      tag: !join [ *VERSION, "-", *DEVICE_TYPE, "-", *TAG_PYTHON_VERSION, "-", *OS_VERSION, "-ec2" ]
+      docker_file: !join [ docker/, *SHORT_VERSION, /, *DOCKER_PYTHON_VERSION, /Dockerfile.graviton.ec2., *DEVICE_TYPE ]
+      context:
+        <<: *INFERENCE_CONTEXT
```
[the remark this answers is not in the document]
Comment (xncqr) [reply tensorflow/buildspec-graviton-2.7.yml]: added

File: tensorflow/inference/docker/2.9/py3/Dockerfile.graviton.cpu [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+ && apt-get update \
+ && apt-get -y install --no-install-recommends \
+    nginx \
+    nginx-module-njs \
+ && apt-get clean \
+ && rm -rf /var/lib/apt/lists/*
+
+RUN ${PIP} install --no-cache-dir \
+    falcon==3.1.0 \
+    gunicorn==20.1.0
```
[the remark this answers is not in the document]
Comment (xncqr) [reply tensorflow/inference/docker/2.9/py3/Dockerfile.graviton.cpu]: Thanks for the catch. I updated the EC2 with the new versions and removed the installing of the packages from the SM image as the packages will already be in place from the EC2

File: testspec.yml (hunk 6 above)
Comment (xncqr) [reply testspec.yml:14]: Comment added. This operation might not last long as we might have something else to use in its place.

File: tensorflow/inference/docker/2.9/py3/Dockerfile.graviton.cpu [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+
+FROM ec2 AS sagemaker
+
+LABEL maintainer="Amazon AI"
+LABEL dlc_major_version="1"
+
+# Specify accept-bind-to-port LABEL for inference pipelines to use SAGEMAKER_BIND_TO_PORT
+# https://docs.aws.amazon.com/sagemaker/latest/dg/inference-pipeline-real-time.html
+LABEL com.amazonaws.sagemaker.capabilities.accept-bind-to-port=true
+LABEL com.amazonaws.sagemaker.capabilities.multi-models=true
```
[the remark this answers is not in the document]
Comment (xncqr) [reply tensorflow/inference/docker/2.9/py3/Dockerfile.graviton.cpu:159]: the endpoint is planned to support multi-model with Graviton.

File: testspec.yml (hunk 6 above)
Comment (tejaschumbalkar) [reply testspec.yml:14]: Is this added explicitly for sanity tests?

**Review by saimidu — COMMENTED** [review]
Buildspec, dockerfile, and test script changes look good.

Please re-enable graviton builds/tests to confirm that the latest changes are correct.

**Review by saimidu — APPROVED** [review]
